### PR TITLE
Implement a reveal_locals expression. (#3387)

### PR DESCRIPTION
**Pull request by aisipos**
This PR is a "rough draft" implementation of a `reveal_locals` expression, which as it stands can produce from this input:

```
def f(a: int, b: int) -> int:
    c = a + b
    reveal_locals()
    return c
```

this output:
```
sample.py:3: error: Revealed local types are '{'a': builtins.int, 'b': builtins.int, 'c': builtins.int}'
```

I am new to the codebase so likely this PR will need a fair bit more work and would like some feedback. I certainly need to add new tests and documentation but wanted to do that after a first round of feedback if possible. 

Some notes:
* Essentially, the implementation is just using the `type_map` attribute of the `TypeChecker` instance to produce the output. I don't know if this is the correct approach but it seemed the simplest to my admittedly cursory knowledge of the codebase.
* I also am storing the `locals` attribute of the `SemanticAnalyzer` into each `RevealLocalsExpr`, but at the moment I am not using it. This could lead to other ways of implementing this but I wasn't sure of the best use of it. 
* Currently the output is on one line, dictionary style. I am open to other possibilities, e.g:
```
sample.py:3: error: Revealed type of 'a': builtins.int
sample.py:3: error: Revealed type of 'b': builtins.int
sample.py:3: error: Revealed type of 'c': builtins.int
```
However, it's worth considering who the "audience" of this output is. The dictionary output is more machine parseable, which could be useful for editor plugins that use this functionality. Admittedly, when I first thought of this feature, this was the use case I was thinking of.
* There is at least one bug in the current implementation:
```
def f(a: int, b: int) -> int:
    reveal_locals()
    c = a + b
    reveal_locals()
    return c
```
gives:
```
sample.py:2: error: Revealed local types are '{}'
sample.py:4: error: Revealed local types are '{'a': builtins.int, 'b': builtins.int, 'c': builtins.int}'
```
The output should be the same for both lines. I do not know of an easy fix yet.

Feedback is welcomed. I realize there will be likely many changes to this before it is merge ready.

### CODE REVIEW

**Comment by aisipos** [comment]
Some discussion with @JukkaL :
as implemented, `reveal_locals()` does show variables from surrounding scopes, which is inconsistent with the name. The choices would be a new name (e.g., `reveal_all` or something), or to actually limit `reveal_locals()` to just local variables. Either is valuable, although perhaps if the primary audience would be editors, it might be useful to have more information than less, so I would lean to keeping the surrounding variables in the output but change the name.

I also fixed unnecessary moving of some classes in b0ca3579e5f5899b46891e2ea4278d71b9753a55

There is still a bug where `reveal_locals()` at the beginning of a function doesn't "see" the function arguments. I'm working to understand why.

**Comment by gvanrossum** [comment]
I prefer just locals; global could be too much.

**Comment by aisipos** [comment]
@ddfisher I fixed some of the issues we discussed. I unified `RevealLocalsExpr` into `RevealTypeExpr`, however in the various traversers I kept the distinction between `visit_reveal_type_node` and `visit_reveal_locals_node`. I did this to avoid having a giant if statement section inside each `visit_reveal_type_node` definition, however if you'd rather I unify the two of them  I'm happy to make that change.

The main downside is I still cannot deal with the issue where:
```
def f(a: int, b: int) -> int:
    reveal_locals()
    c = a + b
    reveal_locals()
    return c
```
The second `return_locals` "sees" c,a,b, but the first `return_locals` sees no variables. I made some experimental commits to try to fix this but reverted them because they were probably a dead end. `reveal_type` is able to see them because it introduces a variable node of the variable it wants to reveal, but since reveal_locals() takes no arguments it doesn't introduce any variable nodes. I still don't understand why the TypeChecker doesn't see the types of the function arguments inside a function, but I will keep looking.

**Comment by aisipos** [comment]
@ddfisher OK, I've rebased this PR against master, added some tests, added some docs, and fixed some test breakage (Travis runs successfully now) . Some of the issues I had with type inference a month ago appear to be gone, I'm assuming changes in the past month have made the method used in this PR work now. Still needs a good review though. The commits are a bit of a mess, but I will squash them once I get some feedback. Thanks for your help.

**Comment by ddfisher** [comment]
Thanks for the updates!  I'll try to take a look in the next couple days!

**Comment by aisipos** [comment]
@ddfisher Due to the recent mypy release, I've rebased this PR on top of master, fixing conflicts and fixing new typing linting errors. All tests are passing again.

**Comment by JukkaL** [comment]
@aisipos Are you interested in finishing this up?

**Comment by aisipos** [comment]
Hi @JukkaL, I would be happy to finish it up. It's been a while since I gave it any attention, but I will try to get it up to date again with master this weekend.

**Comment by aisipos** [comment]
Hi @JukkaL, I have rebased this on top of master and fixed all the tests. However, it should get a review from a maintainer. I'm happy to address any concerns about the functionality or implementation.

File: mypy/treetransform.py [hunk 1]
```diff
@@ -377,8 +377,12 @@ def visit_cast_expr(self, node: CastExpr) -> CastExpr:
         return CastExpr(self.expr(node.expr),
                         self.type(node.type))
 
-    def visit_reveal_type_expr(self, node: RevealTypeExpr) -> RevealTypeExpr:
-        return RevealTypeExpr(self.expr(node.expr))
+    def visit_reveal_expr(self, node: RevealExpr) -> RevealExpr:
+        if node.kind == REVEAL_TYPE:
+            return RevealExpr(kind=REVEAL_TYPE, expr=self.expr(cast(Expression, node.expr)))
```
Comment (JukkaL) [inline mypy/treetransform.py]: Again, I'd use `assert node.expr is not None` here.

File: test-data/unit/check-functions.test [hunk 2]
```diff
@@ -2302,3 +2302,14 @@ f = lambda: 5
 reveal_type(f)
 [out]
 main:2: error: Revealed type is 'def () -> builtins.int'
+
+[case testRevealLocalsFunction]
```
Comment (JukkaL) [inline test-data/unit/check-functions.test]: Thanks for adding tests! I have one idea for additional test: using `reveal_locals()` inside class body.

Comment (aisipos) [reply test-data/unit/check-functions.test]: Is the test `testRevealLocalsClass` sufficient for this?

**Pull request merged**